File: app/src/main/java/org/digitalcampus/oppia/model/Activity.java
/* 
 * This file is part of OppiaMobile - https://digital-campus.org/
 * 
 * OppiaMobile is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * OppiaMobile is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with OppiaMobile. If not, see <http://www.gnu.org/licenses/>.
 */

package org.digitalcampus.oppia.model;

import java.io.File;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;

import org.digitalcampus.mobile.learning.R;
import org.digitalcampus.oppia.exception.GamificationEventNotFound;


public class Activity extends MultiLangInfoModel implements Serializable{
	
	/**
	 * 
	 */
	private static final long serialVersionUID = -1548943805902073988L;

	public static final String TAG = Activity.class.getSimpleName();
	
	private long courseId;
	private int sectionId;
	private int actId;
	private int dbId;
	private String actType;
	private List<Lang> locations = new ArrayList<>();
	private List<Lang> contents = new ArrayList<>();
	private String digest;
	private String imageFile;
	private List<Media> media = new ArrayList<>();
	private boolean completed = false;
	private boolean attempted = false;
	private boolean customImage = false;
	private String mimeType;
	private List<GamificationEvent> gamificationEvents = new ArrayList<>();

	public Activity(){
		// do nothing
	}
	
	public boolean hasCustomImage(){
		return this.customImage;
	}

    public String getImageFilePath(String prefix){
        if(!prefix.endsWith(File.separator)){
            prefix += File.separator;
        }
        return prefix + this.imageFile;
    }

    public int getDefaultResourceImage(){
        if(actType.equals("quiz")){
            return R.drawable.default_icon_quiz;
        } else if (actType.equals("page") && this.hasMedia()){
            return R.drawable.default_icon_video;
        }
        return R.drawable.default_icon_activity;
    }

	public void setImageFile(String imageFile) {
		this.imageFile = imageFile;
		this.customImage = true;
	}
	
	public List<Media> getMedia() {
		return media;
	}

	public Media getMedia(String filename){
		for (Media m : getMedia()) {
			if (m.getFilename().equals(filename)) {
				return m;
			}
		}
		return null;
	}

	public void setMedia(List<Media> media) {
		this.media = media;
	}

	public String getDigest() {
		return digest;
	}

	public void setDigest(String digest) {
		this.digest = digest;
	}
	
	public long getCourseId() {
		return courseId;
	}

	public void setCourseId(long courseId) {
		this.courseId = courseId;
	}

	public int getSectionId() {
		return sectionId;
	}

	public void setSectionId(int sectionId) {
		this.sectionId = sectionId;
	}

	public int getActId() {
		return actId;
	}

	public void setActId(int actId) {
		this.actId = actId;
	}

	public String getActType() {
		return actType;
	}

	public void setActType(String actType) {
		this.actType = actType;
	}

	public String getLocation(String lang) {
		for(Lang l: locations){
			if(l.getLanguage().equals(lang)){
				return l.getContent();
			}
		}
		if(locations.isEmpty()){
			return null;
		} else {
			return locations.get(0).getContent();
		}

	}
	
	public void setLocations(List<Lang> locations) {
		this.locations = locations;
	}
	
	public String getContents(String lang) {
		for(Lang l: contents){
			if(l.getLanguage().equals(lang)){
				return l.getContent();
			}
		}
		if(contents.isEmpty()) {
			return "No content";
		} else {
			return contents.get(0).getContent();
		}

	}
	
	public void setContents(List<Lang> contents) {
		this.contents = contents;
	}

	public boolean hasMedia(){
		return !media.isEmpty();
	}
	
	public void setCompleted(boolean completed){
		this.completed = completed;
	}
	
	public boolean getCompleted(){
		return this.completed;
	}

	public String getMimeType() {
		return mimeType;
	}

	public void setMimeType(String mimeType) {
		this.mimeType = mimeType;
	}

	public boolean isAttempted() {
		return attempted;
	}

	public void setAttempted(boolean attempted) {
		this.attempted = attempted;
	}

	public int getDbId() {
		return dbId;
	}

	public void setDbId(int dbId) {
		this.dbId = dbId;
	}

	public void setGamificationEvents(List<GamificationEvent> events){
		gamificationEvents = events;
	}

    public GamificationEvent findGamificationEvent(String event) throws GamificationEventNotFound {
        for(GamificationEvent ge: gamificationEvents){
            if(ge.getEvent().equals(event)){
                return ge;
            }
        }
        throw new GamificationEventNotFound(event);
    }
}

File: app/src/main/java/org/digitalcampus/oppia/model/Activity.kt
/* 
 * This file is part of OppiaMobile - https://digital-campus.org/
 * 
 * OppiaMobile is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * 
 * OppiaMobile is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with OppiaMobile. If not, see <http://www.gnu.org/licenses/>.
 */
package org.digitalcampus.oppia.model

import android.util.Log
import org.digitalcampus.mobile.learning.R
import org.digitalcampus.oppia.analytics.Analytics
import org.digitalcampus.oppia.exception.GamificationEventNotFound
import org.digitalcampus.oppia.utils.storage.FileUtils
import java.io.File
import java.io.IOException
import java.io.Serializable

class Activity : MultiLangInfoModel(), Serializable {
    companion object {
        private const val serialVersionUID = -1548943805902073988L
        @JvmField
        val TAG = Activity::class.simpleName
    }

    var courseId: Long = 0
    var sectionId = 0
    var actId = 0
    var dbId = 0
    var actType: String? = null
    private var locations: List<Lang> = ArrayList()
    private var contents: List<Lang> = ArrayList()
    var digest: String? = null
    private var imageFile: String? = null
    var media: List<Media> = ArrayList()
    var completed = false
    var isAttempted = false
    private var customImage = false
    var mimeType: String? = null
    private var gamificationEvents: List<GamificationEvent> = ArrayList()
    var wordCount = 0

    fun hasCustomImage(): Boolean {
        return customImage
    }

    fun getImageFilePath(prefix: String): String {
        var prefixedPath = prefix
        if (!prefixedPath.endsWith(File.separator)) {
            prefixedPath += File.separator
        }
        return prefixedPath + imageFile
    }

    fun getDefaultResourceImage(): Int {
        return when(actType) {
            "quiz" -> R.drawable.default_icon_quiz
            "activity" -> if (hasMedia()) R.drawable.default_icon_video else R.drawable.default_icon_activity
            else -> R.drawable.default_icon_activity
        }
    }

    fun setImageFile(imageFile: String?) {
        this.imageFile = imageFile
        customImage = true
    }

    fun hasMedia(): Boolean {
        return media.isNotEmpty()
    }

    fun getMedia(filename: String): Media? {
        return media.find { it.filename == filename }
    }

    fun getLocation(lang: String?): String? {
        val matchingLang = locations.find { it.language.equals(lang, ignoreCase = true) }
        return matchingLang?.content ?: if (locations.isNotEmpty()) locations[0].content else null
    }

    fun setLocations(locations: List<Lang>) {
        this.locations = locations
    }

    fun getContents(lang: String?): String {
        val matchingLang = contents.find { it.language.equals(lang, ignoreCase = true) }
        return matchingLang?.content ?: if (contents.isNotEmpty()) contents[0].content else "No content"
    }

    fun getFileContents(courseLocation: String, lang: String?): String? {
        val fileContent = StringBuilder()
        if (getLocation(lang) != null && actType != "url") {
            val url = courseLocation + getLocation(lang)
            try {
                fileContent.append(" ")
                fileContent.append(FileUtils.readFile(url))
                return fileContent.toString().trim()
            } catch (e: IOException) {
                Analytics.logException(e)
                Log.d(TAG, "IOException:", e)
            }
        }
        return null
    }

    fun setContents(contents: List<Lang>) {
        this.contents = contents
    }

    fun setGamificationEvents(events: List<GamificationEvent>) {
        gamificationEvents = events
    }

    @Throws(GamificationEventNotFound::class)
    fun findGamificationEvent(event: String): GamificationEvent {
        return gamificationEvents.find { it.event == event }
            ?: throw GamificationEventNotFound(event)
    }
}